### ISSUE TITLE
mitogen.core.Router._loop_once() passes sets to select.select()

### ISSUE THREAD

**Issue by moreati** [dw/mitogen#68]
I'm slowly making my way through Mitogen with MyPy. One line it complains about is

https://github.com/dw/mitogen/blob/43fae4414a2103b7ac6d0e09f9d67b777a2aac39/mitogen/core.py#L1016-L1017 

with the message (ignore the line number)

```
mitogen/core.py:1128: error: Argument 1 to "select" has incompatible type "Set[Side]"; expected "Sequence[Any]"
mitogen/core.py:1128: error: Argument 2 to "select" has incompatible type "Set[Side]"; expected "Sequence[Any]"
```
The discrepancy is that `self._readers` & `self._readers` are each a `set` of "waitable objects", i.e. they are *collections*, not *sequences*.

I don't know if could cause undesired behaviour, or it's just MyPy and the [`select.select()`](https://docs.python.org/2/library/select.html#select.select) docs being overly specific. I'll investigate further.

**Comment by dw** [comment]
Hrm annoying, I'd call it a docs bug, since behaviour in CPython (and PyPy) going back forever is that it just takes an iterable. Does MyPy have some kind of override facility?  Converting the sets to lists on every loop would be wasteful, just to match basically a doc bug (the structure passed to select is inherently useless and must internally be iterated to transform it into an fdset anyway).

**Comment by dw** [comment]
We could potentially try just replacing the sets with lists, in the usual case they're only going to contain <5 entries anyway, and the code might even end up faster. The question would be after I get async connect finished, whether scanning the lists for add/removal slows down with a master that has hundreds of connections

**Comment by dw** [comment]
I don't think so.. once a connection is established to a context, it is never removed from the readers set (so there is never any scan of the list until a disconnection occurs).

The writers set only contains contexts that currently have data buffered, and in the typical case it'll simply be empty. I'm not sure about the worst case, though

**Comment by dw** [comment]
mind.. blown.. select() is converting the set into a list internally (via PySequence_Fast()). So there's a real chance swapping the set for a list will improve things

**Comment by moreati** [comment]
THank you for looking. FTR changing the arguments to `sorted(self...)` didn't make any difference to unit test failures.

> Does MyPy have some kind of override facility?

Yes, I can add a `# type: ignore` on a per-line basis

**Comment by moreati** [comment]
FTR I've push the MyPy trial to https://github.com/moreati/mitogen/tree/mypy-try. I'm not proposing to do anything more with it. I think I've learned what I can.

**Comment by dw** [comment]
What's your verdict on MyPy? I'm not completely averse to the idea of it, especially since the annotations are just comments, and don't forget you really did find a bug :).. maybe we could merge it

**Comment by moreati** [comment]
> What's your verdict on MyPy?

On a per bug found basis it was an expensive exercise. As a learning exercise for me it was useful. As a way of documenting the APIs in-situ  I found it very handy. I don't know of a way to feed MyPy annotations into doctools, so there would be some duplicated effort and or chance of discrepancies there.

> I'm not completely averse to the idea of it, especially since the annotations are just comments, and don't forget you really did find a bug :).. maybe we could merge it

Not all the changes are comments, e.g. I sprinked a few asserts, and changed a base class or two

https://github.com/moreati/mitogen/commit/5df74fc5ba9f4043db0605fb9e39727bd1a4b7c8#diff-60eabbf230c5d9137e81bf0f41ed37bbR104
https://github.com/moreati/mitogen/commit/5df74fc5ba9f4043db0605fb9e39727bd1a4b7c8#diff-3540c3617f4566023f000914cebe2d71L317
